packages/components/src/components/combobox/combobox.component.ts

```diff
@@ -0,0 +1,967 @@
+  }
+
+  /** Sets a custom validation message. Pass an empty string to restore validity. */
+  setCustomValidity(message: string) {
+    this.valueInput.setCustomValidity(message);
+    this.formControlController.updateValidity();
+  }
+
+  /** Sets focus on the control. */
+  focus(options?: FocusOptions) {
+    this.displayInput.focus(options);
+  }
+
+  /** Removes focus from the control. */
+  blur() {
+    this.displayInput.blur();
+  }
+
+  private createComboboxOptionsFromQuery(queryString: string) {
+    const optgroups: SynOptGroup[] = [];
+
+    this.filteredOptions = this.getSlottedOptions()
+      .filter(option => this.filter(option, queryString) || queryString === '')
+      .map(option => {
+        const clonedOption = option.cloneNode(true) as SynOption;
+
+        // Check if the option has a syn-optgroup as parent
+        const hasOptgroup = option.parentElement?.tagName.toLowerCase() === 'syn-optgroup';
+        if (!hasOptgroup) {
+          return clonedOption;
+        }
+
+        const optgroup = option.parentElement as SynOptGroup;
+        const filteredOptgroup = optgroups.find((el) => el.id === optgroup.id);
+
+        // Check if the optgroup was already added to the filteredOptions.
+        // It should only be added once!
+        if (filteredOptgroup) {
+          filteredOptgroup?.appendChild(clonedOption);
+          return undefined;
+        }
+
+        const clonedOptgroup = optgroup.cloneNode() as SynOptGroup;
+        clonedOptgroup.appendChild(clonedOption);
+        optgroups.push(clonedOptgroup);
+        return clonedOptgroup;
+      })
+      // we need to remove the undefined values here
+      .filter(el => el) as Array<SynOption | SynOptGroup>;
+  }
+
+  private async handleInput() {
+    const inputValue = this.displayInput.value;
+    this.value = inputValue;
+    await this.updateComplete;
+    this.open = this.filteredWrapper.children.length > 0;
+    this.setSelectedOption(undefined);
+
+    this.formControlController.updateValidity();
+    this.emit('syn-input');
+  }
+
+  private handleChange() {
+    // Only update the value and emit the event, if the change event occurred by
+    // the user typing something in and removing focus of the combobox
+    if (!this.selectedOption) {
+      this.value = this.displayInput.value;
+      // Update validity
+      // eslint-disable-next-line @typescript-eslint/no-floating-promises
+      this.updateComplete.then(() => {
+        this.formControlController.updateValidity();
+      });
+      this.emit('syn-change');
+    }
+  }
+
+  private getSlottedOptions() {
+    return getAllOptions(getAssignedElementsForSlot(this.defaultSlot)).flat();
+  }
+
+  private getSlottedOptGroups(): SynOptGroup[] {
+    return filterOnlyOptgroups(getAssignedElementsForSlot(this.defaultSlot));
+  }
+
+  /* eslint-disable no-param-reassign */
+  private handleDefaultSlotChange() {
+    const slottedOptions = this.getSlottedOptions();
```

Comment:
You could get rid of some parts of the complexity be returning early. It would also make the code more performant
```suggestion
    // Rerun this handler when <syn-option> is registered
    if (!customElements.get('syn-option')) {
      // eslint-disable-next-line @typescript-eslint/no-floating-promises
      customElements.whenDefined('syn-option').then(() => this.handleDefaultSlotChange());
      return;
    }

    const slottedOptions = this.getSlottedOptions();
    const slottedOptgroups = this.getSlottedOptGroups();

    slottedOptions.forEach((option, index) => {
      option.id = option.id || `syn-combobox-option-${index}`;
    });

    slottedOptgroups.forEach((optgroup, index) => {
      optgroup.id = optgroup.id || `syn-combobox-optgroup-${index}`;
    });

    this.createComboboxOptionsFromQuery(this.value);

    if (this.hasFocus && this.value.length > 0 && !this.open) {
      // eslint-disable-next-line @typescript-eslint/no-floating-promises
      this.show();
    }
```